WalletWasabi/WabiSabi/Models/DecompositionAlgs/GreedyDecomposer.cs

```diff
@@ -0,0 +1,67 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.DecompositionAlgs
+{
+	public class GreedyDecomposer
+	{
+		public GreedyDecomposer(IEnumerable<Money> preferredDenominations, Money dustThreshold, FeeRate feeRate)
+		{
+			Decomposition = new(preferredDenominations);
+			DustThreshold = dustThreshold;
+			FeeRate = feeRate;
+		}
+
+		public Decomposition Decomposition { get; }
+		private Money DustThreshold { get; }
+		private FeeRate FeeRate { get; }
+
+		public void Decompose(Coin coin)
+		{
+			Money remaining = coin.Amount - FeeRate.GetFee(coin.ScriptPubKey.EstimateOutputVsize());
+
+			while (remaining > DustThreshold)
+			{
+				Money? denom = LargestDenomBelowIncl(remaining);
+				if (denom is null || denom == Money.Zero)
+				{
+					break;
+				}
+				Decomposition.AddSorted(denom);
+
+				var effectiveCost = coin.Amount + FeeRate.GetFee(coin.ScriptPubKey.EstimateOutputVsize());
+
+				remaining -= effectiveCost;
+			}
+
+			if (remaining > DustThreshold)
+			{
+				Decomposition.AddSorted(remaining);
+			}
+		}
+
+		private Money? LargestDenomBelowIncl(Money amount)
+		{
+			Money result = Decomposition.First();
+			if (result > amount)
+			{
+				return null;
+			}
+
+			foreach (var coin in Decomposition)
+			{
+				if (coin <= amount)
+				{
+					result = coin;
+				}
+				else
+				{
+					break;
+				}
+			}
+			return result;
+		}
```

Comment:
I didn't really understand your previous comment about this method. How is this different that a `FirstOrDefault(x => x.Amount <= amount)`?

Reply:
I don't get it. You would like to replace the whole method with `FirstOrDefault`?

Reply:
Simplified!

Reply:
I would not. But now that you are mentioning...